From 0803bf81bd1417683f69e5c95a984c093672c529 Mon Sep 17 00:00:00 2001
From: Julian Bouzas <julian.bouzas@collabora.com>
Date: Tue, 20 Oct 2026 14:30:22 -0400
Subject: [PATCH] spa: alsa: Don't skip ELD recheck for profiles with input
 mappings

Some devices might have HDMI and analog input on the same codec. This patch
makes sure the ELD is also re-checked for those profiles.
---
 spa/plugins/alsa/acp/alsa-mixer.c | 17 ++++++++++++-----
 1 file changed, 12 insertions(+), 5 deletions(-)

diff --git a/spa/plugins/alsa/acp/alsa-mixer.c b/spa/plugins/alsa/acp/alsa-mixer.c
index 90b52db85..c17a07f75 100644
--- a/spa/plugins/alsa/acp/alsa-mixer.c
+++ b/spa/plugins/alsa/acp/alsa-mixer.c
@@ -5355,14 +5355,11 @@ void pa_alsa_profile_set_recheck_hdmi_eld(
     PA_HASHMAP_FOREACH(p, ps->profiles, state) {
         uint32_t idx;
         bool supports_hdmi_output_mapping = false;
+        bool supports_input_mapping = true;
 
         if (p->profile.flags & (ACP_PROFILE_OFF | ACP_PROFILE_PRO))
             continue;
 
-	/* Skip profiles with input mappings as HDMI ELD is only for playback */
-	if (p->input_mappings)
-	    continue;
-
         if (p->output_mappings) {
             PA_IDXSET_FOREACH(m, p->output_mappings, idx) {
                 /* Only handle mappings for the device whose ELD changed */
@@ -5379,7 +5376,17 @@ void pa_alsa_profile_set_recheck_hdmi_eld(
             }
         }
 
-        if (supports_hdmi_output_mapping) {
+        /* Mappings not affected by the ELD must have been opened during probing */
+        if (p->input_mappings) {
+            PA_IDXSET_FOREACH(m, p->input_mappings, idx) {
+                if (m->supported == 0) {
+                    supports_input_mapping = false;
+                    break;
+                }
+            }
+        }
+
+        if (supports_hdmi_output_mapping && supports_input_mapping) {
             p->supported = true;
             pa_log_debug("HDMI profile %s enabled (ELD channels matched)", p->name);
         }
-- 
2.56.0

